README.md

```diff
@@ -0,0 +1,300 @@
+"every" tree-sitter AST node)
+
+
+
+
+
+##### Writing filter plugins
+
+TODO: add a "guide" for this
+
+The short version is:
+
+While in theory you could probably write filter plugins in other languages the "happy path" would
+be to write them in Rust and use the example filter plugins from [`examples/`](https://github.com/helixbass/tree-sitter-grep/tree/examples)
+as a starting point/reference
+
+The basic idea is that for each tree-sitter AST node that is a potential match according to the supplied
+query argument, the filter plugin then additionally gets invoked and indicates whether it considers that
+node a match or not (basically as a `(&tree_sitter::Node) -> bool` "predicate")
+
+
+
+
+
+#### Supported target languages
+
+Currently, tree-sitter-grep "bakes in" support for searching the following languages:
+
+- C
+- C++
+- C#
+- CSS
+- Dockerfile
+- Elisp
+- Elm
+- Go
+- HTML
+- Java
+- JavaScript
+- JSON
+- Kotlin
+- Lua
+- Objective-C
+- Python
+- Ruby
+- Rust
+- Swift
+- Toml
+- tree-sitter queries (how meta!)
+- TypeScript
+
+In theory, any language that has a tree-sitter grammar crate published/available
+should be "fair game". In the future we may support dynamically specifying/loading
+additional languages
+
+Or feel free to [file an issue](https://github.com/helixbass/tree-sitter-grep/issues)
+requesting "baked-in" support for other languages
+
+
+
+
+#### Restricting the query to specific files/languages
+
+By default, tree-sitter-grep will recursively search all
+"non-ignored/hidden" files of the [supported languages/types](#supported-target-languages)
+and if it can parse the provided query against that language's grammar it will
+then search that file's contents for matches
+
+To explicitly specify/restrict to a single language, use the `-l`/`--language` argument:
+```
+$ tree-sitter-grep -q '(trait_bounds) @t' -l rust
+```
+
+You can also restrict the search to certain files/directories by providing path arguments:
+```
+$ tree-sitter-grep -q '(trait_bounds) @t' src/main.rs src/compiler
+```
+
+
+
+
+#### Additional flags/arguments
+
+For documentation of additional arguments related to eg customizing the match output, run:
+```
+$ tree-sitter-grep --help
+```
+
+In general, we are aiming to be rather [`ripgrep`](https://github.com/BurntSushi/ripgrep)-"compatible"
+
+
+
+
+## Performance
+
+I haven't done any "real" benchmarking but the general take seems to be that tree-sitter-grep
+is pleasantly, surprisingly fast (especially given that tree-sitter is not optimized for the
+"parse-from-scratch" use case)
+
+For "not gigantic" code-bases I'm tending to see it run in < 100ms
+
+And for "gigantic" code-bases where it's eg scanning > 300k lines of code and outputting > 7000 matches,
+I'm seeing it run in say 360ms, which still feels "quite fast"
+
+
+
+
+## Editor integrations
+
+TODO, I believe that @peterstuart has written an initial version of an Emacs plugin
+and I started tinkering with writing a neovim plugin
```

Comment:
It basically works, I should see about packaging it up and submitting it (or at least adding good instructions).